File: 008_string_to_integer_atoi/Solution.py
class FSM:
    def __init__(self):
        self.state = 'start'
        self.sign = 1
        self.ans = 0
        self.table = {
            'start': ['start', 'signed', 'in_number', 'end'],
            'signed': ['end', 'end', 'in_number', 'end'],
            'in_number': ['end', 'end', 'in_number', 'end'],
            'end': ['end', 'end', 'end', 'end'],
        }

    def get_column(self, c: str) -> int:
        if c.isspace():
            return 0
        if c in ['+', '-']:
            return 1
        if c.isdigit():
            return 2
        return 3

    def get(self, c: str):
        self.state = self.table[self.state][self.get_column(c)]
        if self.state == 'in_number':
            self.ans = self.ans * 10 + int(c)
            self.ans = min(self.ans, INT_MAX) if self.sign == 1 else min(self.ans, -INT_MIN)
        elif self.state == 'signed':
            self.sign = 1 if c == '+' else -1


class Solution:
    def myAtoi(self, str: str) -> int:
        fsm = FSM()
        for c in str:
            fsm.get(c)
        return fsm.sign * fsm.ans

File: 008_string_to_integer_atoi/Solution.java
import java.util.HashMap;

class FSM {
    public String status = "start";
    public int sign = 1;
    public long ans = 0;
    public HashMap<String, String[]> table = new HashMap<>();

    public FSM() {
        table.put("start", new String[]{"start", "signed", "in_number", "end"});
        table.put("signed", new String[]{"end", "end", "in_number", "end"});
        table.put("in_number", new String[]{"end", "end", "in_number", "end"});
        table.put("end", new String[]{"end", "end", "end", "end"});
    }

    public int getColumn(char ch) {
        if (ch == ' ') return 0;
        if (ch == '+' || ch == '-') return 1;
        if (ch >= '0' && ch <= '9') return 2;
        return 3;
    }

    public void get(char ch) {
        status = table.get(status)[getColumn(ch)];
        if (status.equals("in_number")) {
            int curInt = Character.getNumericValue(ch);
            ans = ans * 10 + curInt;
            if ((int) ans != ans) {
                if (sign == 1) ans = Integer.MAX_VALUE;
                if (sign == -1) ans = Integer.MIN_VALUE;
            }
        } else if (status.equals("signed")) {
            sign = ch == '+' ? 1 : -1;
        }
    }
}

class Solution {
    public int myAtoi(String str) {
        FSM fsm = new FSM();
        for (char ch : str.toCharArray()) {
            fsm.get(ch);
        }
        return fsm.sign * (int) fsm.ans;
    }
}
